exchanges/coinbasepro/coinbasepro.go

```diff
@@ -1505,6 +1488,65 @@ func (c *CoinbasePro) SendAuthenticatedHTTPRequest(ctx context.Context, ep excha
 	return json.Unmarshal(interim, result)
 }
 
+// GetJWT generates a new JWT
+func (c *CoinbasePro) GetJWT(ctx context.Context, uri string) (string, time.Time, error) {
+	creds, err := c.GetCredentials(ctx)
+	if err != nil {
+		return "", time.Time{}, err
+	}
+	block, _ := pem.Decode([]byte(creds.Secret))
+	if block == nil {
+		return "", time.Time{}, errCantDecodePrivKey
+	}
+	key, err := x509.ParseECPrivateKey(block.Bytes)
+	if err != nil {
+		return "", time.Time{}, err
+	}
+	nonce, err := common.GenerateRandomString(16, "1234567890ABCDEF")
+	if err != nil {
+		return "", time.Time{}, err
+	}
+	regTime := time.Now()
+	mapClaims := jwt.MapClaims{
+		"iss": "cdp",
+		"nbf": regTime.Unix(),
+		"exp": regTime.Add(time.Minute * 2).Unix(),
+		"sub": creds.Key,
+	}
+	if uri != "" {
+		mapClaims["uri"] = uri
+	}
+	tok := jwt.NewWithClaims(jwt.SigningMethodES256, mapClaims)
+	tok.Header["kid"] = creds.Key
+	tok.Header["nonce"] = nonce
+	sign, err := tok.SignedString(key)
+	return sign, regTime, err
+	// The code below mostly works, but seems to lead to bad results on the signature step. Deferring until later
+	// head := map[string]any{"kid": creds.Key, "typ": "JWT", "alg": "ES256", "nonce": nonce}
+	// headJSON, err := json.Marshal(head)
+	// if err != nil {
+	// 	return "", time.Time{}, err
+	// }
+	// headEncode := base64URLEncode(headJSON)
+	// regTime := time.Now()
+	// body := map[string]any{"iss": "cdp", "nbf": regTime.Unix(), "exp": regTime.Add(time.Minute * 2).Unix(), "sub": creds.Key /*, "aud": "retail_rest_api_proxy"*/}
+	// if uri != "" {
+	// 	body["uri"] = uri
+	// }
+	// bodyJSON, err := json.Marshal(body)
+	// if err != nil {
+	// 	return "", time.Time{}, err
+	// }
+	// bodyEncode := base64URLEncode(bodyJSON)
+	// hash := sha256.Sum256([]byte(headEncode + "." + bodyEncode))
+	// sig, err := ecdsa.SignASN1(rand.Reader, key, hash[:])
+	// if err != nil {
+	// 	return "", time.Time{}, err
+	// }
+	// sigEncode := base64URLEncode(sig)
+	// return headEncode + "." + bodyEncode + "." + sigEncode, regTime, nil
```

Comment:
What is this for?

Reply:
As comment at the top of the block says, it's my implementation of the JWT protocol, so we won't have to import a separate package to handle that. Alas, there's some issue I can't identify with generating the signature, which has it create an invalid one, so this can't be used right now.